feat(rust): Add common-hypercache crate for multi-tier caching

haacked opened the pull request:
Implements a `HyperCacheReader` class for reading from the Django `HyperCache`. For now, we only implement reading since that's all we need.

## Problem

For the `local_evaluation` port to Rust, we're going to need code to read from the HyperCache.

## Changes

Adds a new `common-hypercache` crate that implements PostHog's `HyperCache` pattern for reading from multiple cache tiers:
- Redis (primary, fastest)
- S3 (secondary, persistent fallback)
- Automatic Redis backfilling from S3 hits

For now, this implementation only _reads_ from the HyperCache. It doesn't implement writing to the HyperCache.

Also implemented a `common-s3` crate since the HyperCache needs a testable S3 client. The only method implemented is `get_string` as that's all I needed for HyperCache right now.

But I noticed that cymbal and kafka-deduplicator each have a private s3 client implementation. If the authors of those were so inclined, I'd be open to collaborating on making this client support their needs as a follow-up: /cc @oliverb123 and @jose-sequeira 

## How did you test this code?

Unit tests

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

haacked committed:
feat(rust): Add `common-hypercache` crate for multi-tier caching

Adds a new common-hypercache crate that implements PostHog's HyperCache pattern for reading from multiple cache tiers:
- Redis (primary, fastest)
- S3 (secondary, persistent fallback)
- Automatic Redis backfilling from S3 hits

Note: When the HyperCache writes to Redis, it calls json.dumps on the data. That data is pickle encoded and then Zstd compressed if the bytes are more than 512.

So the Rust code has to account for all that.

This also implements a `common-s3` crate. Right now, that only implements what HyperCache needs, but could be used for other Rust services.

haacked committed:
Move common-compression to common/compression

andyzzhao commented on rust/common/hypercache/src/lib.rs:
```diff
@@ -0,0 +1,788 @@
+/// Cache key type matching Django's KeyType = Team | str | int
+#[derive(Debug)]
+pub enum KeyType {
+    Team(Box<dyn TeamIdentifier>),
```
Could this just be `TeamId(TeamId)`?

haacked committed:
Fix spelling

Co-authored-by: Dylan Martin <dylan@posthog.com>

haacked committed:
Add error logging

haacked committed:
feat(rust): Add feature flags and testing infrastructure to S3/HyperCache crates

- Add mock-client feature flag to common-s3 for optional mockall dependency
- Enable proper unit vs integration test separation
- Add cargo-shear ignore for optional dependencies
- Fix integration test imports and naming consistency
- Add feature flag to hypercache to propagate S3 mock support
- Clean up unused dependencies

haacked commented:
Some notes on the latest changes:

## Added a type alias for compile-time client switching:

```rust
#[cfg(not(feature = "mock-client"))]
pub type S3ClientType = S3Impl;      // Production

#[cfg(feature = "mock-client")]
pub type S3ClientType = MockS3Client;     // Tests
```

These rely on the `mock-client` feature flag.

```yaml
[features]
mock-client = ["dep:mockall"]
```

## Why these changes were necessary

**The core problem:** We needed to support both real AWS S3 clients and mock clients for testing, but Rust's native async traits aren't dyn-compatible (can't use `Arc<dyn S3Client>`). The `async fn` syntax in traits prevents trait object creation.

**The solution:** Use compile-time polymorphism instead of runtime polymorphism through feature flags and type aliases.

## What this approach enables

**Clean production builds:**
- Test dependencies (`mockall`) are only included when needed
- Smaller binary size and faster compilation for production
- Zero runtime overhead from abstractions

**Fast development workflow:**

```bash
cargo test --lib          # Unit tests: 0.00s (mocks)
cargo test --test '*'     # Integration tests: 0.21s (real services)
```

**Comprehensive testing strategy:**
- Unit tests use mocks for instant feedback and testing business logic
- Integration tests use real S3/MinIO connections for end-to-end validation
- Both run automatically in CI with full docker-compose environment

## How Integration Tests Work

Integration tests bypass the type alias system entirely and directly use real implementations:

```rust
// Integration tests import real implementations
use common_s3::{S3Impl, S3Client, S3Error};

// Connect to actual MinIO instance
const S3_ENDPOINT: &str = "http://127.0.0.1:19000";
```

This gives us production-like testing without affecting the unit test architecture.

haacked committed:
Upgrade `thiserror` to `2.0`

haacked committed:
Refactor S3 mock to use automock instead of manual implementation

Replace manual MockS3Client implementation with mockall's #[automock]
attribute for better maintainability and idiomatic Rust testing.

Changes:
- Remove manual MockS3Client implementation and builder pattern
- Add #[automock] attribute to S3Client trait
- Update all tests to use automock API with expect_get_string()
- Make mock-client feature default for better developer experience
- Remove unnecessary cargo-shear ignore since mockall is now used
- Remove redundant tests that only verified mockall functionality

Benefits:
- More maintainable mocking with less boilerplate
- Follows Rust testing conventions
- Eliminates unused dependency warnings
- Unit tests work without manual feature flags

haacked committed:
Make django version explicit

Since this is a library, we don't want it to read from the env directly.

haacked commented on rust/common/hypercache/src/lib.rs:
```diff
@@ -0,0 +1,788 @@
+use thiserror::Error;
+use tokio::time::timeout;
+
+/// Metric name for tracking hypercache operations in Prometheus (same one used in Django's HyperCache)
+const HYPERCACHE_COUNTER_NAME: &str = "posthog_hypercache_get_from_cache";
+
+/// Sentinel value used in Redis to indicate that a cache key exists but has no data
+const HYPER_CACHE_EMPTY_VALUE: &str = "__missing__";
```
Ok, I changed the code to allow passing in the Django cache version.

dmarticus replied:
I'm on board

dmarticus commented on rust/Cargo.lock:
```diff
@@ -1703,7 +1703,8 @@ version = "0.1.0"
  "base64 0.22.1",
  "flate2",
- "thiserror 2.0.12",
+ "thiserror 1.0.69",
```
yup, works for m!

dmarticus reviewed (approved):
thanks for responding to the feedback; I like where this is headed :)
